lib/engine/auto_router.rb

```diff
@@ -88,81 +133,328 @@ def compute(corporation, **opts)
           id = chains.flat_map { |c| c[:paths] }.sort!
           next if connections[id]
 
-          connections[id] = chains.map do |c|
+          connection = chains.map do |c|
             { left: c[:nodes][0], right: c[:nodes][1], chain: c }
           end
-        end
-      end
+          connections[id] = connection
 
-      puts "Found #{connections.size} paths in: #{Time.now - now}"
-      puts 'Pruning paths to legal routes'
+          counter += 1
 
-      now = Time.now
-      train_routes = Hash.new { |h, k| h[k] = [] }
-      connections.each do |_, connection|
-        @game.route_trains(corporation).each do |train|
-          route = Engine::Route.new(
-            @game,
-            @game.phase,
-            train,
-            connection_data: connection,
-          )
-          route.revenue
-          train_routes[train] << route
-        rescue GameError # rubocop:disable Lint/SuppressedException
+          # each train has opportunity to vote to abort a branch of this node's path-walk tree
+          path_abort.each { |train, _| path_abort[train] = false }
+
+          # build a test route for each train, use route.revenue to check for errors, keep the good ones
+          trains.each do |train|
+            unless path_abort[train]
+              bitfield = bitfield_from_connection(connection, hexside_bits, bit_to_bitmask)
+              route = Engine::Route.new(
+                @game,
+                @game.phase,
+                train,
+                connection_data: connection,
+                bitfield: bitfield,
+              )
+
+              route.revenue # raises various errors if bad route
+              train_routes[train] << route
+              route_counter[train] += 1
+            end
+
+          # These all result in the route not being added to train_routes[train],
+          # but the nature of the error determines how to continue or terminate processing of the path walk
+          rescue RouteTooLong
+            # ignore for this train, and abort walking this path if ignored for all trains
+            path_abort[train] = true
+            if path_abort.values.all?
+              abort = :abort # this path is dead for all trains, don't walk it further
+            end
+          rescue NoToken, RouteTooShort
+            # keep extending this connection set
+          rescue ReusesCity, RouteBlocked
+            abort = :abort # this path is dead, don't walk it further
+          rescue GameError => e
+            # an unhandled route error that probably should be handled
+            puts e
+          end
+          abort
         end
+        puts ' Node timeout reached' if node_abort
       end
-      puts "Pruned paths to #{train_routes.map { |k, v| k.name + ':' + v.size.to_s }.join(', ')} in: #{Time.now - now}"
+
+      # Check that there are no duplicate hexside bits (algorithm error)
+      mismatch = hexside_bits.length - hexside_bits.uniq.length
```

Comment:
don't use length, use size

Reply:
why is this even necessary? why would the hexside bits not be unique? you made hexside bits yourself